Objects/frameobject.c

```diff
@@ -243,7 +243,19 @@ framelocalsproxy_setitem(PyObject *self, PyObject *key, PyObject *value)
             Py_XINCREF(value);
             PyCell_SetTakeRef((PyCellObject *)cell, value);
         } else if (value != PyStackRef_AsPyObjectBorrow(oldvalue)) {
-            PyStackRef_XCLOSE(fast[i]);
+            if (!PyStackRef_IsNull(fast[i])) {
+                if (frame->f_overwritten_fast_locals == NULL) {
+                    frame->f_overwritten_fast_locals = PyList_New(0);
```

Comment:
It feels like it would be safer to use tuples here, since there are technically ways of finding this list and clearing it.

Reply:
Is that possible to do from "normal" (i.e. without using something like ctypes) Python code? There are lots of ways to crash the vm if you're determined enough. IMHO using a list is fine, so long as you can't clear it without concerted effort.

Reply:
I was thinking something like `gc.get_referents(frame)` or `gc.get_referrers(thing_i_want_to_free)`. Not a hill I'm willing to die on, but a tuple just seems safer here if the intended purpose is to guarantee that the object won't be decrefed out of existence. Especially since this is a rare, slow case where people are already doing some dubious introspective stuff.

Reply:
Tuples are also (typically) tracked by the GC so I'm not sure it helps with things like `gc.get_referrers()`.

I don't think we can untrack this either -- we want to be able to collect the frame if there is a reference cycle between something in `f_overwritten_fast_locals` and the frame.

Reply:
Oh, I just realized that the important consideration is that tuples are immutable...

Reply:
@brandtbucher - Makes sense. I've updated the PR to use a tuple instead. Thanks for the suggestion!